### PR TITLE
[WIP] Issue76 apo nice

### LINKED ISSUE #76: Allow users with APO roles access to the APO itself
A person with any role in an APO should have access to view the APO. 

A person with a dor-apo-manager role in an APO should be able to edit the APO object itself as well as create a collection that is governed by that APO. 

Example Jon Manton in the ARS APO https://argo.stanford.edu/catalog/druid:bz845pv2292

### PR DESCRIPTION
**Pull request by darrenleeweber**
This fixes #76 

This is blocked by https://github.com/sul-dlss/argo/pull/563, which will replace the "Consolidate Argo user groups in the User model" (commit 347e7fd8f1fd2f2b02788c9bed83b3f7cc2e1a66).  That PR also has some overlap with the work in "General review of user roles and groups" (commit 653b702868a9ec8939ce47276e5b7ce09b68cd33, mostly code style/efficiency changes).  @jmartin-sul and I have previously reviewed most of those changes and they might get pulled out of this PR.

The core of this PR is in `User#can_admin?`, `User#can_manage?` and `User#can_view?`
(commit 02e697c50439ff0c50f2a8d2926a47b4d2af672c).  These public methods call a very important private method.  The essence of the permission refactoring is in the new logic in `User#request_object_permission`.  It should be possible to make `User#roles` a private method also, but it is left public and the further analysis and discussion required for that can be another PR.

The use of these methods in the application is in "Apply User#can_admin?, User#can_manage? and User#can_view?" (commit ea1b53b43831ebc006588dcc253d13a20d380cb3).

The last part of this PR is the use of the new User model in specs.  The main part of this is to create methods in the spec_helper and use them consistently in specs that require a certain type of user (mostly `admin_user`).  All of this is in "Revise specs after application of User#can_admin?, User#can_manage? ..." (commit 676f8b48b0efdbf844ef8d004cebe5b6c0675d7a).  Part of this work motivated changes for more consistent `id` parsing in the `find_druid` work, has been isolated in https://github.com/sul-dlss/argo/pull/555.  So the WIP on this PR is trying to separate the changes from those, otherwise this PR is blocked by #555 because the commits there need to be disentangled from those in this PR.

### CODE REVIEW

**Comment by darrenleeweber** [comment]
Blocked by https://github.com/sul-dlss/argo/pull/532
Blocked by https://github.com/sul-dlss/argo/pull/543

File: app/controllers/registration_controller.rb [hunk 1]
```diff
@@ -59,44 +63,48 @@ def collection_list
...
+        # TODO: this is stupid, should handle "stanford" regardless of the string's case, but the xpath parser doesn't support the lower-case() fn
+        # readable by stanford translates to Stanford
+        # TODO: found something indicating that xpath might support regex
+        default_opt = 'stanford'
+      elsif adm_xml.xpath('//rightsMetadata/access[@type=\'read\']/machine/none').length > 0
+        # readable by none is either Citation Only (formerly "None") or Dark
+        if adm_xml.xpath('//rightsMetadata/access[@type=\'discover\']/machine/world').length > 0
+          # discoverable by world but readable by none translates to Citation Only/none
+          default_opt = 'none'
+        elsif adm_xml.xpath('//rightsMetadata/access[@type=\'discover\']/machine/none').length > 0
```
Comment (darrenleeweber) [inline app/controllers/registration_controller.rb]: This is addressed by https://github.com/sul-dlss/argo/pull/543

File: app/controllers/registration_controller.rb [hunk 2]
```diff
@@ -59,44 +63,48 @@ def collection_list
...
-      elsif adm_xml.xpath('//rightsMetadata/access[@type=\'discover\']/machine/none').length > 0
-        # discoverable by none and readable by none translates to Dark
-        default_opt = 'dark'
+      # FIXME: should not be in Controller
+      # figure out what the default option (if any) should be
+      default_opt = nil
+      if adm_xml.xpath('//rightsMetadata/access[@type=\'read\']/machine/world').length > 0
+        # readable by world translates to World
+        default_opt = 'world'
+      elsif adm_xml.xpath('//rightsMetadata/access[@type=\'read\']/machine/group[text()=\'Stanford\' or text()=\'stanford\']').length > 0
```
Comment (darrenleeweber) [inline app/controllers/registration_controller.rb]: This is addressed by https://github.com/sul-dlss/argo/pull/543

File: app/controllers/registration_controller.rb [hunk 3]
```diff
@@ -59,44 +63,48 @@ def collection_list
...
+        # TODO: this is stupid, should handle "stanford" regardless of the string's case, but the xpath parser doesn't support the lower-case() fn
+        # readable by stanford translates to Stanford
+        # TODO: found something indicating that xpath might support regex
+        default_opt = 'stanford'
+      elsif adm_xml.xpath('//rightsMetadata/access[@type=\'read\']/machine/none').length > 0
```
Comment (darrenleeweber) [inline app/controllers/registration_controller.rb]: This is addressed by https://github.com/sul-dlss/argo/pull/543

File: app/controllers/registration_controller.rb [hunk 4]
```diff
@@ -59,44 +63,48 @@ def collection_list
...
+        # readable by none is either Citation Only (formerly "None") or Dark
+        if adm_xml.xpath('//rightsMetadata/access[@type=\'discover\']/machine/world').length > 0
```
Comment (darrenleeweber) [inline app/controllers/registration_controller.rb]: This is addressed by https://github.com/sul-dlss/argo/pull/543

File: app/controllers/registration_controller.rb [hunk 5]
```diff
@@ -59,44 +63,48 @@ def collection_list
...
-      if adm_xml.xpath('//rightsMetadata/access[@type=\'discover\']/machine/world').length > 0
-        # discoverable by world but readable by none translates to Citation Only/none
-        default_opt = 'none'
-      elsif adm_xml.xpath('//rightsMetadata/access[@type=\'discover\']/machine/none').length > 0
-        # discoverable by none and readable by none translates to Dark
-        default_opt = 'dark'
+      # FIXME: should not be in Controller
+      # figure out what the default option (if any) should be
+      default_opt = nil
+      if adm_xml.xpath('//rightsMetadata/access[@type=\'read\']/machine/world').length > 0
```
Comment (darrenleeweber) [inline app/controllers/registration_controller.rb]: This is addressed by https://github.com/sul-dlss/argo/pull/543

**Comment by tingulfsen** [comment]
As far as I can tell the changes in this pull request work well for solving the issue in #334 and so from that perspective, this is ready to be merged.

File: app/controllers/items_controller.rb [hunk 6]
```diff
@@ -690,9 +687,9 @@ def add_workflow
 
     # We need to sync up the workflows datastream with workflow service (using #find)
     # and then force a committed Solr update before redirection.
-    reindex Dor::Item.find(params[:id])
-    msg = "Added #{wf_name}"
+    reindex find_druid(params[:id])
```
Comment (drh-stanford) [inline app/controllers/items_controller.rb]: `Dor::Item.find` and `Dor.find` do not follow the same logic, so calling `find_druid` (which uses `Dor.find`) in this case changes the behavior.

**Comment by darrenleeweber** [comment]
Blocked by #555

File: app/controllers/items_controller.rb (hunk 6 above)
Comment (darrenleeweber) [inline app/controllers/items_controller.rb]: To be addressed in #555

Comment (darrenleeweber) [inline app/controllers/items_controller.rb]: Deleting and migrating these comments to #555.

**Comment by darrenleeweber** [comment]
Closing this because the code branches are evolving too fast for this to keep up.  The details of this work need to be evaluated in smaller chunks.

**Pull request closed without merging**